Added Linter and Formatter plugins

iCharlesHu opened the pull request:
This PR adds the missing pieces needed for linter and formatter. It includes:
- `FoundationPreviewLinter`: linter command plugin that lints the entire repo. Can be invoked via `swift package lint`
- `FoundationPreviewFormatter`: formatter command line plugin. Can be invoked via `swift package format` (you will need to grant it permission to write to the package repository)
- `Hooks/pre-push`: a `pre-push` that automatically lint the repo upon `git push`.

---
The linter rules is configured via `linter-rules`. Right now it's **mostly** consists of whitespace and line break rules. My plan is to add each additional rules in its own PR for better feedbacks.

To reduce the risk of merge conflicts, this PR does not contain the actual format changes. I'll push a separate PR after this one is merged.

iCharlesHu commented:
@swift-ci please test

iCharlesHu commented on linter-rules.json:
```diff
@@ -0,0 +1,55 @@
+{
+    "version": 1,
+    "lineLength": 9999,
```
Most Swift repos don't enforce max line width... we should prob do the same.

iCharlesHu commented on linter-rules.json:
```diff
@@ -0,0 +1,55 @@
+        "DontRepeatTypeInStaticProperties": true,
+        "FileScopedDeclarationPrivacy": true,
+        "FullyIndirectEnum": true,
+        "GroupNumericLiterals": true,
```
This one could be controversial. (It basically formats numeric literals like `10_000_000`. I turned it on because it's mostly automatically and it does look nice.

iCharlesHu commented on linter-rules.json:
```diff
@@ -0,0 +1,55 @@
+        "UseSynthesizedInitializer": true,
+        "UseTripleSlashForDocumentationComments": true,
+        "UseWhereClausesInForLoops": false,
+        "ValidateDocumentationComments": false,
```
Hopefully we could turn this on one day 😊

jmschonfeld commented:
Between
> a pre-push that automatically lint the repo upon git push.
and
> To reduce the risk of merge conflicts, this PR does not contain the actual format changes. I'll push a separate PR after this one is merged.

Does that mean that once this is merged, any future pushes will be blocked until your change with the format goes in?

Also, do we have a mechanism to perform linting in CI to call out any issues that might slip through?

iCharlesHu commented:
> 
> Does that mean that once this is merged, any future pushes will be blocked until your change with the format goes in?
> 
No. You'll have to manually "install" the hook by copying it to `.git/hooks/`. Only then will it automatically perform linting upon `git push`.

>
> Also, do we have a mechanism to perform linting in CI to call out any issues that might slip through?
>
We should. I'll ask @shahmishal to add `swift package lint` to the pipeline.

jmschonfeld commented on Package.swift:
```diff
@@ -18,7 +18,10 @@ let package = Package(
+            revision: "0c1de7149a39a9ff82d4db66234dec587b30a3ad"),
+        .package(
+            url: "https://github.com/apple/swift-format.git",
+            branch: "main")
```
Do we want to tie ourselves to an unstable branch here, or should we add a release-based dependency?

jmschonfeld commented on Plugins/FoundationPreviewLinter/main.swift:
```diff
@@ -0,0 +1,47 @@
+        let configFile = context.package.directory.appending(["linter-rules.json"])
+
+        for target in context.package.targets {
+            guard let sourceTarget = target as? SourceModuleTarget else {
```
Does this include test targets?

iCharlesHu commented:
@swift-ci please test

jmschonfeld commented on linter-rules.json:
```diff
@@ -0,0 +1,55 @@
+        "NeverForceUnwrap": false,
+        "NeverUseForceTry": false,
+        "NeverUseImplicitlyUnwrappedOptionals": false,
+        "NoAccessLevelOnExtensionDeclaration": false,
```
I know @lorentey has opinions on this one, not sure what other packages do but this seems like one that might be worthwhile turning on 🙂

iCharlesHu commented:
Major changes since first commit:
- Switched to https://github.com/nicklockwood/SwiftFormat since it's more feature complete.
- Updated the `linter-rules` to only include the "strictly necessary" (or "non controversial") rules that mainly deal with spaces and line breaks.  My plan is to introduce each additional rules in its own PR for better feedbacks.

iCharlesHu commented:
Same as before, I will introduce the actual format changes in a different PR.
Example diff:
```diff
diff --git a/Sources/FoundationEssentials/AttributedString/AttributeContainer.swift b/Sources/FoundationEssentials/AttributedString/AttributeContainer.swift
index a336dc4..945992e 100644
--- a/Sources/FoundationEssentials/AttributedString/AttributeContainer.swift
+++ b/Sources/FoundationEssentials/AttributedString/AttributeContainer.swift
@@ -18,13 +18,13 @@ package import _RopeModule
 
 @dynamicMemberLookup
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
-public struct AttributeContainer : Sendable {
-    internal var storage : AttributedString._AttributeStorage
-    
+public struct AttributeContainer: Sendable {
+    internal var storage: AttributedString._AttributeStorage
+
     public init() {
         storage = .init()
     }
-    
+
     internal init(_ storage: AttributedString._AttributeStorage) {
         self.storage = storage
     }
@@ -33,13 +33,13 @@ public struct AttributeContainer : Sendable {
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
 extension AttributeContainer {
     @preconcurrency
-    public subscript<T: AttributedStringKey>(_: T.Type) -> T.Value? where T.Value : Sendable {
+    public subscript<T: AttributedStringKey>(_: T.Type) -> T.Value? where T.Value: Sendable {
         get { storage[T.self] }
         set { storage[T.self] = newValue }
     }
 
     @preconcurrency
-    public subscript<K: AttributedStringKey>(dynamicMember keyPath: KeyPath<AttributeDynamicLookup, K>) -> K.Value? where K.Value : Sendable {
+    public subscript<K: AttributedStringKey>(dynamicMember keyPath: KeyPath<AttributeDynamicLookup, K>) -> K.Value? where K.Value: Sendable {
         get { self[K.self] }
         set { self[K.self] = newValue }
     }
@@ -70,11 +70,11 @@ extension AttributeContainer {
         return Builder(container: self)
     }
 
-    public struct Builder<T: AttributedStringKey> : Sendable {
-        var container : AttributeContainer
+    public struct Builder<T: AttributedStringKey>: Sendable {
+        var container: AttributeContainer
 
         @preconcurrency
-        public func callAsFunction(_ value: T.Value) -> AttributeContainer where T.Value : Sendable {
+        public func callAsFunction(_ value: T.Value) -> AttributeContainer where T.Value: Sendable {
             var new = container
             new[T.self] = value
             return new
@@ -85,9 +85,9 @@ extension AttributeContainer {
         self.storage.mergeIn(other.storage, mergePolicy: mergePolicy)
     }
 
-    public func merging(_ other: AttributeContainer, mergePolicy:  AttributedString.AttributeMergePolicy = .keepNew) -> AttributeContainer {
+    public func merging(_ other: AttributeContainer, mergePolicy: AttributedString.AttributeMergePolicy = .keepNew) -> AttributeContainer {
         var copy = self
-        copy.merge(other, mergePolicy:  mergePolicy)
+        copy.merge(other, mergePolicy: mergePolicy)
         return copy
     }
 }
diff --git a/Sources/FoundationEssentials/AttributedString/AttributeScope.swift b/Sources/FoundationEssentials/AttributedString/AttributeScope.swift
index a4ffc96..66633b1 100644
--- a/Sources/FoundationEssentials/AttributedString/AttributeScope.swift
+++ b/Sources/FoundationEssentials/AttributedString/AttributeScope.swift
@@ -17,7 +17,7 @@
 // }
 // An AttributeScope can contain other scopes as well.
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
-public protocol AttributeScope : DecodingConfigurationProviding, EncodingConfigurationProviding {
+public protocol AttributeScope: DecodingConfigurationProviding, EncodingConfigurationProviding {
     static var decodingConfiguration: AttributeScopeCodableConfiguration { get }
     static var encodingConfiguration: AttributeScopeCodableConfiguration { get }
 }
@@ -25,18 +25,18 @@ public protocol AttributeScope : DecodingConfigurationProviding, EncodingConfigu
 @_nonSendable
 @frozen
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
-public enum AttributeScopes { }
+public enum AttributeScopes {}
 
 #if FOUNDATION_FRAMEWORK
 
 import Darwin
 @_implementationOnly import ReflectionInternal
 
-fileprivate struct ScopeDescription : Sendable {
-    var attributes: [String : any AttributedStringKey.Type] = [:]
-    var markdownAttributes: [String : any MarkdownDecodableAttributedStringKey.Type] = [:]
+fileprivate struct ScopeDescription: Sendable {
+    var attributes: [String: any AttributedStringKey.Type] = [:]
+    var markdownAttributes: [String: any MarkdownDecodableAttributedStringKey.Type] = [:]
     var subscopes: [any AttributeScope.Type] = []
-    
+
     mutating func merge(_ other: Self) {
         attributes.merge(other.attributes, uniquingKeysWith: { current, new in new })
         markdownAttributes.merge(other.markdownAttributes, uniquingKeysWith: { current, new in new })
@@ -44,17 +44,17 @@ fileprivate struct ScopeDescription : Sendable {
     }
 }
 
-fileprivate struct LoadedScopeCache : Sendable {
-    private var scopeMangledNames : [String : (any AttributeScope.Type)?]
+fileprivate struct LoadedScopeCache: Sendable {
+    private var scopeMangledNames: [String: (any AttributeScope.Type)?]
     private var lastImageCount: UInt32
-    private var scopeContents : [Type : ScopeDescription]
-    
+    private var scopeContents: [Type: ScopeDescription]
+
     init() {
         scopeMangledNames = [:]
         scopeContents = [:]
         lastImageCount = 0
     }
-    
+
     mutating func scopeType(
         for name: String,
         in path: String
@@ -75,18 +75,18 @@ fileprivate struct LoadedScopeCache : Sendable {
                 $0.value != nil
             }
         }
-        
+
         guard let handle = dlopen(path, RTLD_NOLOAD),
-             let symbol = dlsym(handle, name) else {
+              let symbol = dlsym(handle, name) else {
             scopeMangledNames[name] = nil
             return nil
         }
-        
+
         let type = unsafeBitCast(symbol, to: Any.Type.self) as? any AttributeScope.Type
         scopeMangledNames[name] = type
         return type
     }
-    
+
     subscript(_ type: any AttributeScope.Type) -> ScopeDescription? {
         get {
             scopeContents[Type(type)]
@@ -99,7 +99,7 @@ fileprivate struct LoadedScopeCache : Sendable {
 
 fileprivate let _loadedScopeCache = LockedState(initialState: LoadedScopeCache())
 
-internal func _loadDefaultAttributes() -> [String : any AttributedStringKey.Type] {
+internal func _loadDefaultAttributes() -> [String: any AttributedStringKey.Type] {
     // On native macOS, the UI framework that gets loaded is AppKit. On
     // macCatalyst however, we load a version of UIKit.
     #if !targetEnvironment(macCatalyst)
@@ -134,12 +134,12 @@ internal func _loadDefaultAttributes() -> [String : any AttributedStringKey.Type
             (
                 "$s10Foundation15AttributeScopesO13AccessibilityE0D10AttributesVN",
                 "/System/Library/Frameworks/Accessibility.framework/Accessibility"
-            )
+            ),
         ].compactMap {
             cache.scopeType(for: $0.0, in: $0.1)
         }
     }
-    
+
     // Walk each scope (checking the cache) and gather each scope's attribute table
     let defaultAttributeTypes = (defaultScopes + [AttributeScopes.FoundationAttributes.self]).map {
         $0.attributeKeyTypes()
@@ -158,7 +158,7 @@ internal extension AttributeScope {
         if let cached = _loadedScopeCache.withLock({ $0[Self.self] }) {
             return cached
         }
-        
+
         var desc = ScopeDescription()
         for field in Type(Self.self).fields {
             switch field.type.swiftType {
@@ -178,15 +178,15 @@ internal extension AttributeScope {
         }
         return desc
     }
-    
-    static func attributeKeyTypes() -> [String : any AttributedStringKey.Type] {
+
+    static func attributeKeyTypes() -> [String: any AttributedStringKey.Type] {
         let description = Self.scopeDescription
         return description.subscopes.reduce(description.attributes) {
             $0.merging($1.attributeKeyTypes(), uniquingKeysWith: { current, new in new })
         }
     }
-    
-    static func markdownKeyTypes() -> [String : any MarkdownDecodableAttributedStringKey.Type] {
+
+    static func markdownKeyTypes() -> [String: any MarkdownDecodableAttributedStringKey.Type] {
         let description = Self.scopeDescription
         return description.subscopes.reduce(description.markdownAttributes) {
             $0.merging($1.markdownKeyTypes(), uniquingKeysWith: { current, new in new })
diff --git a/Sources/FoundationEssentials/AttributedString/AttributedString+AttributeTransformation.swift b/Sources/FoundationEssentials/AttributedString/AttributedString+AttributeTransformation.swift
index ea4c049..deb3017 100644
--- a/Sources/FoundationEssentials/AttributedString/AttributedString+AttributeTransformation.swift
+++ b/Sources/FoundationEssentials/AttributedString/AttributedString+AttributeTransformation.swift
@@ -19,11 +19,11 @@ package import _RopeModule
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
 extension AttributedString {
     @preconcurrency
-    public struct SingleAttributeTransformer<T: AttributedStringKey> : Sendable where T.Value : Sendable {
+    public struct SingleAttributeTransformer<T: AttributedStringKey>: Sendable where T.Value: Sendable {
         public var range: Range<Index>
 
         internal var attrName = T.name
-        internal var attr : _AttributeValue?
+        internal var attr: _AttributeValue?
 
         public var value: T.Value? {
             get { attr?.rawValue(as: T.self) }
@@ -31,13 +31,13 @@ extension AttributedString {
         }
 
         @preconcurrency
-        public mutating func replace<U: AttributedStringKey>(with key: U.Type, value: U.Value) where U.Value : Sendable {
+        public mutating func replace<U: AttributedStringKey>(with key: U.Type, value: U.Value) where U.Value: Sendable {
             attrName = key.name
             attr = .init(value, for: U.self)
         }
 
         @preconcurrency
-        public mutating func replace<U: AttributedStringKey>(with keyPath: KeyPath<AttributeDynamicLookup, U>, value: U.Value) where U.Value : Sendable {
+        public mutating func replace<U: AttributedStringKey>(with keyPath: KeyPath<AttributeDynamicLookup, U>, value: U.Value) where U.Value: Sendable {
             self.replace(with: U.self, value: value)
         }
     }
@@ -77,7 +77,7 @@ extension AttributedString {
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
 extension AttributedString {
     public func transformingAttributes<K>(
-        _ k:  K.Type,
+        _ k: K.Type,
         _ c: (inout AttributedString.SingleAttributeTransformer<K>) -> Void
     ) -> AttributedString {
         let orig = AttributedString(_guts)
@@ -94,7 +94,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2>(
-        _ k:  K1.Type,
+        _ k: K1.Type,
         _ k2: K2.Type,
         _ c: (inout AttributedString.SingleAttributeTransformer<K1>,
               inout AttributedString.SingleAttributeTransformer<K2>) -> Void
@@ -117,7 +117,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2, K3>(
-        _ k:  K1.Type,
+        _ k: K1.Type,
         _ k2: K2.Type,
         _ k3: K3.Type,
         _ c: (inout AttributedString.SingleAttributeTransformer<K1>,
@@ -146,7 +146,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2, K3, K4>(
-        _ k:  K1.Type,
+        _ k: K1.Type,
         _ k2: K2.Type,
         _ k3: K3.Type,
         _ k4: K4.Type,
@@ -181,7 +181,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2, K3, K4, K5>(
-        _ k:  K1.Type,
+        _ k: K1.Type,
         _ k2: K2.Type,
         _ k3: K3.Type,
         _ k4: K4.Type,
@@ -232,7 +232,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2>(
-        _ k:  KeyPath<AttributeDynamicLookup, K1>,
+        _ k: KeyPath<AttributeDynamicLookup, K1>,
         _ k2: KeyPath<AttributeDynamicLookup, K2>,
         _ c: (inout AttributedString.SingleAttributeTransformer<K1>,
               inout AttributedString.SingleAttributeTransformer<K2>) -> Void
@@ -241,7 +241,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2, K3>(
-        _ k:  KeyPath<AttributeDynamicLookup, K1>,
+        _ k: KeyPath<AttributeDynamicLookup, K1>,
         _ k2: KeyPath<AttributeDynamicLookup, K2>,
         _ k3: KeyPath<AttributeDynamicLookup, K3>,
         _ c: (inout AttributedString.SingleAttributeTransformer<K1>,
@@ -252,7 +252,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2, K3, K4>(
-        _ k:  KeyPath<AttributeDynamicLookup, K1>,
+        _ k: KeyPath<AttributeDynamicLookup, K1>,
         _ k2: KeyPath<AttributeDynamicLookup, K2>,
         _ k3: KeyPath<AttributeDynamicLookup, K3>,
         _ k4: KeyPath<AttributeDynamicLookup, K4>,
@@ -265,7 +265,7 @@ extension AttributedString {
     }
 
     public func transformingAttributes<K1, K2, K3, K4, K5>(
-        _ k:  KeyPath<AttributeDynamicLookup, K1>,
+        _ k: KeyPath<AttributeDynamicLookup, K1>,
         _ k2: KeyPath<AttributeDynamicLookup, K2>,
         _ k3: KeyPath<AttributeDynamicLookup, K3>,
         _ k4: KeyPath<AttributeDynamicLookup, K4>,
diff --git a/Sources/FoundationEssentials/AttributedString/AttributedString+CharacterView.swift b/Sources/FoundationEssentials/AttributedString/AttributedString+CharacterView.swift
index ad1caa5..44d52de 100644
--- a/Sources/FoundationEssentials/AttributedString/AttributedString+CharacterView.swift
+++ b/Sources/FoundationEssentials/AttributedString/AttributedString+CharacterView.swift
@@ -18,7 +18,7 @@ package import _RopeModule
 
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
 extension AttributedString {
-    public struct CharacterView : Sendable {
+    public struct CharacterView: Sendable {
         /// The guts of the base attributed string.
         internal var _guts: Guts
 
@@ -47,7 +47,7 @@ extension AttributedString {
             let substring = _guts.string[range]
             _range = Range(uncheckedBounds: (substring.startIndex, substring.endIndex))
         }
-        
+
         public init() {
             self.init(Guts())
         }
@@ -111,8 +111,7 @@ extension AttributedString.CharacterView: BidirectionalCollection {
         .init(_range.upperBound)
     }
 
-    @_alwaysEmitIntoClient
-    public var count: Int {
+    @_alwaysEmitIntoClient public var count: Int {
         if #available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) {
             return _count
         }
@@ -120,8 +119,7 @@ extension AttributedString.CharacterView: BidirectionalCollection {
     }
 
     @available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
-    @usableFromInline
-    internal var _count: Int {
+    @usableFromInline internal var _count: Int {
         _characters.count
     }
 
@@ -218,7 +216,7 @@ extension AttributedString.CharacterView: BidirectionalCollection {
             self._replaceSubrange(i ..< j, with: String(newValue))
         }
     }
-    
+
     // Note: This subscript returning a Slice is a bug; unfortunately, this is ABI.
     public subscript(bounds: Range<AttributedString.Index>) -> Slice<AttributedString.CharacterView> {
         get {
@@ -282,9 +280,9 @@ extension AttributedString.CharacterView: RangeReplaceableCollection {
         precondition(
             subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex,
             "AttributedString index range out of bounds")
-        
+
         let subrange = _guts.characterRange(roundingDown: subrange._bstringRange)
-        
+
         // Prevent the BigString mutation below from falling back to Character-by-Character loops.
         if let newElements = _specializingCast(newElements, to: Self.self) {
             _replaceSubrange(subrange, with: newElements._characters)
diff --git a/Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift b/Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift
index 6d9ed46..884f9fd 100644
--- a/Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift
+++ b/Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift
@@ -17,7 +17,7 @@ package import _RopeModule
 #endif
 
 extension AttributedString {
-    internal final class Guts : @unchecked Sendable {
+    internal final class Guts: @unchecked Sendable {
         typealias Index = AttributedString.Index
         typealias Runs = AttributedString.Runs
         typealias AttributeMergePolicy = AttributedString.AttributeMergePolicy
@@ -109,8 +109,8 @@ extension AttributedString.Guts {
 
         var it1 = left.makeIterator()
         var it2 = right.makeIterator()
-    loop:
-        while true {
+        loop:
+            while true {
             switch (it1.next(), it2.next()) {
             case let (leftRun?, rightRun?):
                 guard leftRun.attributes == rightRun.attributes else { return false }
@@ -353,7 +353,7 @@ extension AttributedString.Guts {
 
     func setAttributeValue<K: AttributedStringKey>(
         _ value: K.Value, forKey key: K.Type, in range: Range<BigString.Index>
-    ) where K.Value : Sendable {
+    ) where K.Value: Sendable {
         let value = _AttributeValue(value, for: K.self)
         self.setAttributeValue(value, forKey: K.name, in: range)
     }
diff --git a/Sources/FoundationEssentials/AttributedString/AttributedString+Runs+AttributeSlices.swift b/Sources/FoundationEssentials/AttributedString/AttributedString+Runs+AttributeSlices.swift
index 42b8fb9..9bcdbbe 100644
--- a/Sources/FoundationEssentials/AttributedString/AttributedString+Runs+AttributeSlices.swift
+++ b/Sources/FoundationEssentials/AttributedString/AttributedString+Runs+AttributeSlices.swift
@@ -19,8 +19,8 @@ package import _RopeModule
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
 extension AttributedString.Runs {
     @preconcurrency
-    public struct AttributesSlice1<T : AttributedStringKey> : BidirectionalCollection, Sendable
-    where T.Value : Sendable {
+    public struct AttributesSlice1<T: AttributedStringKey>: BidirectionalCollection, Sendable
+        where T.Value: Sendable {
         public typealias Index = AttributedString.Index
 
         // FIXME: Why no labels?
@@ -106,11 +106,11 @@ extension AttributedString.Runs {
         // down to the nearest valid indices.
     }
 
-    public subscript<T : AttributedStringKey>(_ keyPath: KeyPath<AttributeDynamicLookup, T>) -> AttributesSlice1<T> {
+    public subscript<T: AttributedStringKey>(_ keyPath: KeyPath<AttributeDynamicLookup, T>) -> AttributesSlice1<T> {
         return AttributesSlice1<T>(runs: self)
     }
 
-    public subscript<T : AttributedStringKey>(_ t: T.Type) -> AttributesSlice1<T> {
+    public subscript<T: AttributedStringKey>(_ t: T.Type) -> AttributesSlice1<T> {
         return AttributesSlice1<T>(runs: self)
     }
 }
@@ -119,13 +119,12 @@ extension AttributedString.Runs {
 extension AttributedString.Runs {
     @preconcurrency
     public struct AttributesSlice2<
-        T : AttributedStringKey,
-        U : AttributedStringKey
-    > : BidirectionalCollection, Sendable
-    where
-        T.Value : Sendable,
-        U.Value : Sendable
-    {
+        T: AttributedStringKey,
+        U: AttributedStringKey
+    >: BidirectionalCollection, Sendable
+        where
+        T.Value: Sendable,
+        U.Value: Sendable {
         public typealias Index = AttributedString.Index
 
         // FIXME: Why no labels?
@@ -133,7 +132,7 @@ extension AttributedString.Runs {
 
         internal typealias Runs = AttributedString.Runs
 
-        let runs : Runs
+        let runs: Runs
         let _names: [String]
         let _constraints: [AttributeRunBoundaries]
 
@@ -212,9 +211,9 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey
+    >(
         _ t: KeyPath<AttributeDynamicLookup, T>,
         _ u: KeyPath<AttributeDynamicLookup, U>
     ) -> AttributesSlice2<T, U> {
@@ -222,9 +221,9 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey
+    >(
         _ t: T.Type,
         _ u: U.Type
     ) -> AttributesSlice2<T, U> {
@@ -236,15 +235,14 @@ extension AttributedString.Runs {
 extension AttributedString.Runs {
     @preconcurrency
     public struct AttributesSlice3<
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey
-    > : BidirectionalCollection, Sendable
-    where
-        T.Value : Sendable,
-        U.Value : Sendable,
-        V.Value : Sendable
-    {
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey
+    >: BidirectionalCollection, Sendable
+        where
+        T.Value: Sendable,
+        U.Value: Sendable,
+        V.Value: Sendable {
         public typealias Index = AttributedString.Index
 
         // FIXME: Why no labels?
@@ -252,7 +250,7 @@ extension AttributedString.Runs {
 
         internal typealias Runs = AttributedString.Runs
 
-        let runs : Runs
+        let runs: Runs
         let _names: [String]
         let _constraints: [AttributeRunBoundaries]
 
@@ -335,10 +333,10 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey
+    >(
         _ t: KeyPath<AttributeDynamicLookup, T>,
         _ u: KeyPath<AttributeDynamicLookup, U>,
         _ v: KeyPath<AttributeDynamicLookup, V>
@@ -347,10 +345,10 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey
+    >(
         _ t: T.Type,
         _ u: U.Type,
         _ v: V.Type
@@ -363,17 +361,16 @@ extension AttributedString.Runs {
 extension AttributedString.Runs {
     @preconcurrency
     public struct AttributesSlice4<
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey,
-        W : AttributedStringKey
-    > : BidirectionalCollection, Sendable
-    where
-        T.Value : Sendable,
-        U.Value : Sendable,
-        V.Value : Sendable,
-        W.Value : Sendable
-    {
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey,
+        W: AttributedStringKey
+    >: BidirectionalCollection, Sendable
+        where
+        T.Value: Sendable,
+        U.Value: Sendable,
+        V.Value: Sendable,
+        W.Value: Sendable {
         public typealias Index = AttributedString.Index
 
         // FIXME: Why no labels?
@@ -381,7 +378,7 @@ extension AttributedString.Runs {
 
         internal typealias Runs = AttributedString.Runs
 
-        let runs : Runs
+        let runs: Runs
         let _names: [String]
         let _constraints: [AttributeRunBoundaries]
 
@@ -471,11 +468,11 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey,
-        W : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey,
+        W: AttributedStringKey
+    >(
         _ t: KeyPath<AttributeDynamicLookup, T>,
         _ u: KeyPath<AttributeDynamicLookup, U>,
         _ v: KeyPath<AttributeDynamicLookup, V>,
@@ -485,11 +482,11 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey,
-        W : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey,
+        W: AttributedStringKey
+    >(
         _ t: T.Type,
         _ u: U.Type,
         _ v: V.Type,
@@ -503,19 +500,18 @@ extension AttributedString.Runs {
 extension AttributedString.Runs {
     @preconcurrency
     public struct AttributesSlice5<
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey,
-        W : AttributedStringKey,
-        X : AttributedStringKey
-    > : BidirectionalCollection, Sendable
-    where
-        T.Value : Sendable,
-        U.Value : Sendable,
-        V.Value : Sendable,
-        W.Value : Sendable,
-        X.Value : Sendable
-    {
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey,
+        W: AttributedStringKey,
+        X: AttributedStringKey
+    >: BidirectionalCollection, Sendable
+        where
+        T.Value: Sendable,
+        U.Value: Sendable,
+        V.Value: Sendable,
+        W.Value: Sendable,
+        X.Value: Sendable {
         public typealias Index = AttributedString.Index
 
         // FIXME: Why no labels?
@@ -523,7 +519,7 @@ extension AttributedString.Runs {
 
         internal typealias Runs = AttributedString.Runs
 
-        let runs : Runs
+        let runs: Runs
         let _names: [String]
         let _constraints: [AttributeRunBoundaries]
 
@@ -617,12 +613,12 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey,
-        W : AttributedStringKey,
-        X : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey,
+        W: AttributedStringKey,
+        X: AttributedStringKey
+    >(
         _ t: KeyPath<AttributeDynamicLookup, T>,
         _ u: KeyPath<AttributeDynamicLookup, U>,
         _ v: KeyPath<AttributeDynamicLookup, V>,
@@ -633,12 +629,12 @@ extension AttributedString.Runs {
     }
 
     public subscript <
-        T : AttributedStringKey,
-        U : AttributedStringKey,
-        V : AttributedStringKey,
-        W : AttributedStringKey,
-        X : AttributedStringKey
-    > (
+        T: AttributedStringKey,
+        U: AttributedStringKey,
+        V: AttributedStringKey,
+        W: AttributedStringKey,
+        X: AttributedStringKey
+    >(
         _ t: T.Type,
         _ u: U.Type,
         _ v: V.Type,
@@ -654,7 +650,7 @@ extension AttributedString.Runs {
 @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
 extension AttributedString.Runs {
     @_spi(AttributedString)
-    public struct NSAttributesSlice : BidirectionalCollection, Sendable {
+    public struct NSAttributesSlice: BidirectionalCollection, Sendable {
         public typealias Index = AttributedString.Index
 
         // FIXME: Why no labels?
@@ -819,4 +815,3 @@ extension RangeReplaceableCollection {
         if let item5 { self.append(item5) }
     }
 }
```

iCharlesHu commented:
@swift-ci  please test
